byzcoin/statetrie_test.go

```diff
@@ -71,3 +82,125 @@ func TestStateTrie(t *testing.T) {
+// Use with
+//   go test -v -cpuprofile cpu.prof -run TestDarcRetrieval
+//   go tool pprof -pdf cpu.prof
+func TestDarcRetrieval(t *testing.T) {
+	darcDepth := 3
+	darcWidth := 10
+	entries := 10000
+
+	s := newSer(t, 1, 10*time.Second)
+	defer s.local.CloseAll()
+	s.local.Check = onet.CheckNone
+
+	st, err := s.service().getStateTrie(s.genesis.SkipChainID())
+	require.NoError(t, err)
+	require.NotNil(t, st)
+	require.NotEqual(t, -1, st.GetIndex())
+
+	log.Lvl1("Creating random entries")
+	value := random.Bits(1024, true, random.New())
+	start := time.Now()
+	var scs []StateChange
+	for e := 1; e <= entries; e++ {
+		scs = append(scs, StateChange{
+			StateAction: Create,
+			InstanceID:  random.Bits(256, true, random.New()),
+			ContractID:  ContractDarcID,
+			Value:       value,
+		})
+		if e%100 == 0 {
+			st.StoreAll(scs, 5, CurrentVersion)
+			scs = []StateChange{}
+			log.Print(e, time.Now().Sub(start))
+			start = time.Now()
+		}
+	}
+
+	log.Lvl1("Creating darcs")
+	counter := big.NewInt(0)
+	// Store the tree of darcs in a 2D-array with the root in index 0 and
+	// the leafs in index darcDepth-1
+	darcs := make([][]*darc.Darc, darcDepth)
+	for d := darcDepth - 1; d >= 0; d-- {
+		width := math.Exp(big.NewInt(int64(darcWidth)),
+			big.NewInt(int64(d))).Int64()
+		darcs[d] = make([]*darc.Darc, width)
+		// The leafs just hold a dummy reference to a darc
+		if d == darcDepth-1 {
+			for i := range darcs[d] {
+				counter.Add(counter, big.NewInt(1))
+				id := make([]byte, 32)
+				copy(id, counter.Bytes())
+				ids := []darc.Identity{darc.NewIdentityDarc(id)}
+				rules := darc.InitRules(ids, ids)
+				darcs[d][i] = darc.NewDarc(rules, []byte(counter.String()))
+			}
+		} else {
+			for i := range darcs[d] {
+				counter.Add(counter, big.NewInt(1))
+				ids := make([]darc.Identity, darcWidth)
+				for id := range ids {
+					ids[id] = darc.NewIdentityDarc(
+						darcs[d+1][i*darcWidth+id].GetBaseID())
+				}
+				rules := darc.InitRules(ids, ids)
+				darcs[d][i] = darc.NewDarc(rules, []byte(counter.String()))
+			}
+		}
+	}
+
+	scs = make([]StateChange, 0, counter.Int64())
+	for d := range darcs {
+		for w := range darcs[d] {
+			td := darcs[d][w]
+			log.Lvlf3("darc[%d][%d]: %x has %s", d, w,
+				td.GetBaseID(), td.Rules)
+			buf, err := td.ToProto()
+			require.NoError(t, err)
+			scs = append(scs, StateChange{
+				StateAction: Create,
+				InstanceID:  td.GetBaseID(),
+				ContractID:  ContractDarcID,
+				Value:       buf,
+				Version:     0,
+			})
+		}
+	}
+	require.NoError(t, st.StoreAll(scs, 5, CurrentVersion))
+
+	getDarcs := func(s string, latest bool) *darc.Darc {
+		if !latest {
+			log.Error("cannot handle intermediate darcs")
+			return nil
+		}
+		id, err := hex.DecodeString(strings.Replace(s, "darc:", "", 1))
+		if err != nil || len(id) != 32 {
+			log.Error("invalid darc id", s, len(id), err)
+			return nil
+		}
+		d, err := st.LoadDarcFromTrie(id)
+		if err != nil {
+			return nil
```

Comment:
No `log.Error` (not that I approve of this way to mark error but yeah, otherwise, we would need to change `darc.GetDarc` which might be out of scope for this PR (but not of https://github.com/dedis/cothority/issues/2260!))

Reply:
okay, that makes sense.